### DISCUSSION TITLE
BOSH Support

### DISCUSSION THREAD

**Discussion by kimimaru4000** [vitalyster/SharpXMPP#141] [General]
Hi, does SharpXMPP support the BOSH ([Bi-directional streams Over Synchronous HTTP](https://xmpp.org/extensions/xep-0124.html)) XMPP extension? One use for this is Open Streaming Platform (OSP)'s chat, [which uses XMPP/BOSH](https://wiki.openstreamingplatform.com/en/Install/Chat), and it would awesome to be able to use this library to connect a bot to the chat. Thank you!

**Comment by vitalyster** [comment]
It uses ejabberd, so you can switch from BOSH to Websocket with minimal config changes. Websocket connection is supported by this library.

**Reply by kimimaru4000** [reply]
Hi, I'm looking into this more now.

1. Is it possible to connect over BOSH with this library if the server does not support WebSockets? I want to clarify that the OSP example is just one use-case I'm accounting for, thus I would not always be the server operator.
2. How can I join an MUC room with WebSockets? `XmppWebSocketConnection` does not have a `BookmarkManager`. Attempting to join the MUC chat room directly doesn't work as a result. One-to-one chats work just fine.

Thank you very much!

**Reply by vitalyster** [reply]
1. BOSH is not supported and there is no any reason to support it, it is deprecated by WebSockets
2. `XmppWebSocketConnection` is the `XmppConnection` subclass, so `BookmarkManager` should work the same way

**Reply by kimimaru4000** [reply]
Thank you for the quick reply!

1. Thanks for the clarity on the state of BOSH!
2. Unfortunately, `BookmarkManager` is defined only in [`XmppClient`](https://github.com/vitalyster/SharpXMPP/blob/master/SharpXMPP.Shared/XmppClient.cs). Should it be moved to [`XmppConnection`](https://github.com/vitalyster/SharpXMPP/blob/master/SharpXMPP.Shared/XmppConnection.cs) instead?

**Reply by vitalyster** [reply]
2. Maybe, but you can create one for now in your code as a temporary solution?

**Reply by kimimaru4000** [reply]
Yes, that should work! I'd love to see it in `XmppConnection` to help with cleaner code (Ex. polymorphism) if it seems appropriate to add in there.

I appreciate the help!

**Reply by vitalyster** [reply]
Patches are welcome! :)
